qutip/entropy.py

```diff
@@ -244,24 +243,42 @@ def _entropy_relative(rho, sigma, base=e, sparse=False):
     rel_ent : float
         Value of relative entropy.
 
+    References
+    ----------
+
+    See Nielsen & Chuang, "Quantum Computation and Quantum Information",
+    Section 11.3.1, pg. 511 for a detailed explanation of quantum relative
+    entropy.
     """
-    if rho.type != 'oper' or sigma.type != 'oper':
-        raise TypeError("Inputs must be density matrices..")
-    # sigma terms
-    svals = sp_eigs(sigma.data, sigma.isherm, vecs=False, sparse=sparse)
-    snzvals = svals[svals != 0]
+    if not rho.isoper or not sigma.isoper:
+        raise TypeError("Inputs must be density matrices.")
     if base == 2:
-        slogvals = log2(snzvals)
+        log_base = log2
     elif base == e:
-        slogvals = log(snzvals)
+        log_base = log
     else:
         raise ValueError("Base must be 2 or e.")
-    # rho terms
-    rvals = sp_eigs(rho.data, rho.isherm, vecs=False, sparse=sparse)
-    rnzvals = rvals[rvals != 0]
-    # calculate tr(rho*log sigma)
-    rel_trace = float(real(sum(rnzvals * slogvals)))
-    return -entropy_vn(rho, base, sparse) - rel_trace
+    # S(rho || sigma) = sum_i(p_i log p_i) - sum_ij(p_i P_ij log q_i)
+    #
+    # S is +inf if the kernel of sigma (i.e. svecs[svals == 0]) has non-trivial
+    # intersection with the support of rho (i.e. rvecs[rvals != 0]).
+    rvals, rvecs = sp_eigs(rho.data, rho.isherm, vecs=True, sparse=sparse)
+    svals, svecs = sp_eigs(sigma.data, sigma.isherm, vecs=True, sparse=sparse)
+    nzrvals = rvals[rvals != 0]
```

Comment:
This is a tricky problem numerically - `eigh` isn't guaranteed to find 0 eigenvalues to be _exactly_ 0, so you may need to take a tolerance argument, and compare `rvals[np.abs(rvals) < tol]` to get the nonzeros.

Reply:
I just checked the relative entropy function and ran into this issue. The relative entropy between |0> and |+> should be +inf. However, I got a value of ~53.25 because the svals of |+> returned [-8.71576399e-33  1.00000000e+00] instead of [0 1]. 
Formally the function seems correct to me (implementing eq. 11.53 in Nielsen & Chuang). Thank you very much for implementing it!

Reply:
I'll add a tolerance argument. This was on my mind to add too.

Reply:
Added.